File: src/backend/nfse/src/model/nfse_detalhe_model.py
from src import db
from src.model.nfse_lista_servico_model import NfseListaServicoModel


class NfseDetalheModel(db.Model):
    __tablename__ = 'nfse_detalhe'

    id = db.Column(db.Integer, primary_key=True)
    codigo_cnae = db.Column(db.String(7))
    codigo_tributacao_municipio = db.Column(db.String(20))
    valor_servicos = db.Column(db.Float)
    valor_deducoes = db.Column(db.Float)
    valor_pis = db.Column(db.Float)
    valor_cofins = db.Column(db.Float)
    valor_inss = db.Column(db.Float)
    valor_ir = db.Column(db.Float)
    valor_csll = db.Column(db.Float)
    valor_base_calculo = db.Column(db.Float)
    aliquota = db.Column(db.Float)
    valor_iss = db.Column(db.Float)
    valor_liquido = db.Column(db.Float)
    outras_retencoes = db.Column(db.Float)
    valor_credito = db.Column(db.Float)
    iss_retido = db.Column(db.String(1))
    valor_iss_retido = db.Column(db.Float)
    valor_desconto_condicionado = db.Column(db.Float)
    valor_desconto_incondicionado = db.Column(db.Float)
    municipio_prestacao = db.Column(db.Integer)
    discriminacao = db.Column(db.Text)
    id_nfse_cabecalho = db.Column(db.Integer, db.ForeignKey('nfse_cabecalho.id'))
    id_nfse_lista_servico = db.Column(db.Integer, db.ForeignKey('nfse_lista_servico.id'))

    nfse_lista_servico_model = db.relationship('NfseListaServicoModel', foreign_keys=[id_nfse_lista_servico])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfse_cabecalho = data.get('idNfseCabecalho')
        self.id_nfse_lista_servico = data.get('idNfseListaServico')
        self.codigo_cnae = data.get('codigoCnae')
        self.codigo_tributacao_municipio = data.get('codigoTributacaoMunicipio')
        self.valor_servicos = data.get('valorServicos')
        self.valor_deducoes = data.get('valorDeducoes')
        self.valor_pis = data.get('valorPis')
        self.valor_cofins = data.get('valorCofins')
        self.valor_inss = data.get('valorInss')
        self.valor_ir = data.get('valorIr')
        self.valor_csll = data.get('valorCsll')
        self.valor_base_calculo = data.get('valorBaseCalculo')
        self.aliquota = data.get('aliquota')
        self.valor_iss = data.get('valorIss')
        self.valor_liquido = data.get('valorLiquido')
        self.outras_retencoes = data.get('outrasRetencoes')
        self.valor_credito = data.get('valorCredito')
        self.iss_retido = data.get('issRetido')
        self.valor_iss_retido = data.get('valorIssRetido')
        self.valor_desconto_condicionado = data.get('valorDescontoCondicionado')
        self.valor_desconto_incondicionado = data.get('valorDescontoIncondicionado')
        self.municipio_prestacao = data.get('municipioPrestacao')
        self.discriminacao = data.get('discriminacao')

    def serialize(self):
        return {
            'id': self.id,
            'idNfseCabecalho': self.id_nfse_cabecalho,
            'idNfseListaServico': self.id_nfse_lista_servico,
            'codigoCnae': self.codigo_cnae,
            'codigoTributacaoMunicipio': self.codigo_tributacao_municipio,
            'valorServicos': self.valor_servicos,
            'valorDeducoes': self.valor_deducoes,
            'valorPis': self.valor_pis,
            'valorCofins': self.valor_cofins,
            'valorInss': self.valor_inss,
            'valorIr': self.valor_ir,
            'valorCsll': self.valor_csll,
            'valorBaseCalculo': self.valor_base_calculo,
            'aliquota': self.aliquota,
            'valorIss': self.valor_iss,
            'valorLiquido': self.valor_liquido,
            'outrasRetencoes': self.outras_retencoes,
            'valorCredito': self.valor_credito,
            'issRetido': self.iss_retido,
            'valorIssRetido': self.valor_iss_retido,
            'valorDescontoCondicionado': self.valor_desconto_condicionado,
            'valorDescontoIncondicionado': self.valor_desconto_incondicionado,
            'municipioPrestacao': self.municipio_prestacao,
            'discriminacao': self.discriminacao,
            'nfseListaServicoModel': self.nfse_lista_servico_model.serialize() if self.nfse_lista_servico_model else None,
        }

File: src/frontend/nfse/lib/app/data/model/nfse_detalhe_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfse/app/data/model/model_imports.dart';
import 'package:nfse/app/data/domain/domain_imports.dart';

class NfseDetalheModel {
	int? id;
	int? idNfseCabecalho;
	int? idNfseListaServico;
	String? codigoCnae;
	String? codigoTributacaoMunicipio;
	double? valorServicos;
	double? valorDeducoes;
	double? valorPis;
	double? valorCofins;
	double? valorInss;
	double? valorIr;
	double? valorCsll;
	double? valorBaseCalculo;
	double? aliquota;
	double? valorIss;
	double? valorLiquido;
	double? outrasRetencoes;
	double? valorCredito;
	String? issRetido;
	double? valorIssRetido;
	double? valorDescontoCondicionado;
	double? valorDescontoIncondicionado;
	int? municipioPrestacao;
	String? discriminacao;
	NfseListaServicoModel? nfseListaServicoModel;

	NfseDetalheModel({
		this.id,
		this.idNfseCabecalho,
		this.idNfseListaServico,
		this.codigoCnae,
		this.codigoTributacaoMunicipio,
		this.valorServicos,
		this.valorDeducoes,
		this.valorPis,
		this.valorCofins,
		this.valorInss,
		this.valorIr,
		this.valorCsll,
		this.valorBaseCalculo,
		this.aliquota,
		this.valorIss,
		this.valorLiquido,
		this.outrasRetencoes,
		this.valorCredito,
		this.issRetido,
		this.valorIssRetido,
		this.valorDescontoCondicionado,
		this.valorDescontoIncondicionado,
		this.municipioPrestacao,
		this.discriminacao,
		this.nfseListaServicoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo_cnae',
		'codigo_nfse_municipio',
		'valor_servicos',
		'valor_deducoes',
		'valor_pis',
		'valor_cofins',
		'valor_inss',
		'valor_ir',
		'valor_csll',
		'valor_base_calculo',
		'aliquota',
		'valor_iss',
		'valor_liquido',
		'outras_retencoes',
		'valor_credito',
		'iss_retido',
		'valor_iss_retido',
		'valor_desconto_condicionado',
		'valor_desconto_incondicionado',
		'municipio_prestacao',
		'discriminacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo Cnae',
		'Codigo Tributacao Municipio',
		'Valor Servicos',
		'Valor Deducoes',
		'Valor Pis',
		'Valor Cofins',
		'Valor Inss',
		'Valor Ir',
		'Valor Csll',
		'Valor Base Calculo',
		'Aliquota',
		'Valor Iss',
		'Valor Liquido',
		'Outras Retencoes',
		'Valor Credito',
		'Iss Retido',
		'Valor Iss Retido',
		'Valor Desconto Condicionado',
		'Valor Desconto Incondicionado',
		'Municipio Prestacao',
		'Discriminacao',
	];

	NfseDetalheModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfseCabecalho = jsonData['idNfseCabecalho'];
		idNfseListaServico = jsonData['idNfseListaServico'];
		codigoCnae = jsonData['codigoCnae'];
		codigoTributacaoMunicipio = jsonData['codigoTributacaoMunicipio'];
		valorServicos = jsonData['valorServicos']?.toDouble();
		valorDeducoes = jsonData['valorDeducoes']?.toDouble();
		valorPis = jsonData['valorPis']?.toDouble();
		valorCofins = jsonData['valorCofins']?.toDouble();
		valorInss = jsonData['valorInss']?.toDouble();
		valorIr = jsonData['valorIr']?.toDouble();
		valorCsll = jsonData['valorCsll']?.toDouble();
		valorBaseCalculo = jsonData['valorBaseCalculo']?.toDouble();
		aliquota = jsonData['aliquota']?.toDouble();
		valorIss = jsonData['valorIss']?.toDouble();
		valorLiquido = jsonData['valorLiquido']?.toDouble();
		outrasRetencoes = jsonData['outrasRetencoes']?.toDouble();
		valorCredito = jsonData['valorCredito']?.toDouble();
		issRetido = NfseDetalheDomain.getIssRetido(jsonData['issRetido']);
		valorIssRetido = jsonData['valorIssRetido']?.toDouble();
		valorDescontoCondicionado = jsonData['valorDescontoCondicionado']?.toDouble();
		valorDescontoIncondicionado = jsonData['valorDescontoIncondicionado']?.toDouble();
		municipioPrestacao = jsonData['municipioPrestacao'];
		discriminacao = jsonData['discriminacao'];
		nfseListaServicoModel = jsonData['nfseListaServicoModel'] == null ? NfseListaServicoModel() : NfseListaServicoModel.fromJson(jsonData['nfseListaServicoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfseCabecalho'] = idNfseCabecalho != 0 ? idNfseCabecalho : null;
		jsonData['idNfseListaServico'] = idNfseListaServico != 0 ? idNfseListaServico : null;
		jsonData['codigoCnae'] = codigoCnae;
		jsonData['codigoTributacaoMunicipio'] = codigoTributacaoMunicipio;
		jsonData['valorServicos'] = valorServicos;
		jsonData['valorDeducoes'] = valorDeducoes;
		jsonData['valorPis'] = valorPis;
		jsonData['valorCofins'] = valorCofins;
		jsonData['valorInss'] = valorInss;
		jsonData['valorIr'] = valorIr;
		jsonData['valorCsll'] = valorCsll;
		jsonData['valorBaseCalculo'] = valorBaseCalculo;
		jsonData['aliquota'] = aliquota;
		jsonData['valorIss'] = valorIss;
		jsonData['valorLiquido'] = valorLiquido;
		jsonData['outrasRetencoes'] = outrasRetencoes;
		jsonData['valorCredito'] = valorCredito;
		jsonData['issRetido'] = NfseDetalheDomain.setIssRetido(issRetido);
		jsonData['valorIssRetido'] = valorIssRetido;
		jsonData['valorDescontoCondicionado'] = valorDescontoCondicionado;
		jsonData['valorDescontoIncondicionado'] = valorDescontoIncondicionado;
		jsonData['municipioPrestacao'] = municipioPrestacao;
		jsonData['discriminacao'] = discriminacao;
		jsonData['nfseListaServicoModel'] = nfseListaServicoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfseCabecalho = plutoRow.cells['idNfseCabecalho']?.value;
		idNfseListaServico = plutoRow.cells['idNfseListaServico']?.value;
		codigoCnae = plutoRow.cells['codigoCnae']?.value;
		codigoTributacaoMunicipio = plutoRow.cells['codigoTributacaoMunicipio']?.value;
		valorServicos = plutoRow.cells['valorServicos']?.value?.toDouble();
		valorDeducoes = plutoRow.cells['valorDeducoes']?.value?.toDouble();
		valorPis = plutoRow.cells['valorPis']?.value?.toDouble();
		valorCofins = plutoRow.cells['valorCofins']?.value?.toDouble();
		valorInss = plutoRow.cells['valorInss']?.value?.toDouble();
		valorIr = plutoRow.cells['valorIr']?.value?.toDouble();
		valorCsll = plutoRow.cells['valorCsll']?.value?.toDouble();
		valorBaseCalculo = plutoRow.cells['valorBaseCalculo']?.value?.toDouble();
		aliquota = plutoRow.cells['aliquota']?.value?.toDouble();
		valorIss = plutoRow.cells['valorIss']?.value?.toDouble();
		valorLiquido = plutoRow.cells['valorLiquido']?.value?.toDouble();
		outrasRetencoes = plutoRow.cells['outrasRetencoes']?.value?.toDouble();
		valorCredito = plutoRow.cells['valorCredito']?.value?.toDouble();
		issRetido = plutoRow.cells['issRetido']?.value != '' ? plutoRow.cells['issRetido']?.value : 'S';
		valorIssRetido = plutoRow.cells['valorIssRetido']?.value?.toDouble();
		valorDescontoCondicionado = plutoRow.cells['valorDescontoCondicionado']?.value?.toDouble();
		valorDescontoIncondicionado = plutoRow.cells['valorDescontoIncondicionado']?.value?.toDouble();
		municipioPrestacao = plutoRow.cells['municipioPrestacao']?.value;
		discriminacao = plutoRow.cells['discriminacao']?.value;
		nfseListaServicoModel = NfseListaServicoModel();
		nfseListaServicoModel?.descricao = plutoRow.cells['nfseListaServicoModel']?.value;
	}	

	NfseDetalheModel clone() {
		return NfseDetalheModel(
			id: id,
			idNfseCabecalho: idNfseCabecalho,
			idNfseListaServico: idNfseListaServico,
			codigoCnae: codigoCnae,
			codigoTributacaoMunicipio: codigoTributacaoMunicipio,
			valorServicos: valorServicos,
			valorDeducoes: valorDeducoes,
			valorPis: valorPis,
			valorCofins: valorCofins,
			valorInss: valorInss,
			valorIr: valorIr,
			valorCsll: valorCsll,
			valorBaseCalculo: valorBaseCalculo,
			aliquota: aliquota,
			valorIss: valorIss,
			valorLiquido: valorLiquido,
			outrasRetencoes: outrasRetencoes,
			valorCredito: valorCredito,
			issRetido: issRetido,
			valorIssRetido: valorIssRetido,
			valorDescontoCondicionado: valorDescontoCondicionado,
			valorDescontoIncondicionado: valorDescontoIncondicionado,
			municipioPrestacao: municipioPrestacao,
			discriminacao: discriminacao,
		);			
	}

	
}